internal/v1/handler_blueprints.go

```diff
@@ -277,6 +320,78 @@ func (h *Handlers) UpdateBlueprint(ctx echo.Context, blueprintId uuid.UUID) erro
+			Errors: []HTTPError{{
+				Title:  "Invalid blueprint name",
+				Detail: blueprintInvalidNameDetail,
+			}},
+		})
+	}
+
+	if blueprintRequest.Customizations.Users != nil {
+		for i := range *blueprintRequest.Customizations.Users {
+			requestUser := (*blueprintRequest.Customizations.Users)[i]
+
+			// Make sure every user has either ssh key or password set
+			passwordAndSSHKeyEmptyStringRequest := (requestUser.Password != nil && len(*requestUser.Password) == 0) && (requestUser.SshKey != nil && len(*requestUser.SshKey) == 0)
+			if passwordAndSSHKeyEmptyStringRequest {
+				return ctx.JSON(http.StatusUnprocessableEntity, HTTPErrorList{
+					Errors: []HTTPError{{
+						Title:  "Invalid user",
+						Detail: fmt.Sprintf("User ('%s') must have either a password or an SSH key set.", requestUser.Name),
+					}},
+				})
+			}
+
+			missingPasswordOrSSHKey := !requestUser.hasPassword() || !requestUser.hasSshKey()
+			if missingPasswordOrSSHKey {
+				blueprintEntry, err := h.server.db.GetBlueprint(ctx.Request().Context(), blueprintId, userID.OrgID(), nil)
+				if err != nil {
+					return err
+				}
+				existingUsersMap, err := getExistingUsersMapFromLatestBlueprint(blueprintEntry)
+				if err != nil {
+					return err
+				}
+
+				//Try to use the existing password/ssh_key from the latest blueprint version
+				if existingUser, ok := existingUsersMap[requestUser.Name]; ok {
+					// Set values from current version only if they are not being set in the request (those have precedence)
+					if existingUser.hasPassword() && requestUser.Password == nil {
+						(*blueprintRequest.Customizations.Users)[i].setPassword(existingUser.Password)
+					}
+					if existingUser.hasSshKey() && requestUser.SshKey == nil {
+						(*blueprintRequest.Customizations.Users)[i].setSshKey(existingUser.SshKey)
+					}
+
+					// This means we were able to find password or ssh_key in previous version of blueprint
+					hasPassword := (*blueprintRequest.Customizations.Users)[i].hasPassword()
+					hasSSHKey := (*blueprintRequest.Customizations.Users)[i].hasSshKey()
+					if !hasPassword && !hasSSHKey {
+						return ctx.JSON(http.StatusUnprocessableEntity, HTTPErrorList{
+							Errors: []HTTPError{{
+								Title: "Invalid user",
+								Detail: fmt.Sprintf(
+									"Can't use empty string as a value because user ('%s') must have either a password or an SSH key set.",
+									requestUser.Name),
+							}},
+						})
+					}
+				}
+
+				// If we are here, it means that user has one of password or ssh_key set to non-empty value
+				// If there is empty string in password or ssh_key, it means that we should remove it (set to nil)
+				if requestUser.Password != nil && len(*requestUser.Password) == 0 {
+					(*blueprintRequest.Customizations.Users)[i].RedactPassword()
+				}
+				if requestUser.SshKey != nil && len(*requestUser.SshKey) == 0 {
+					(*blueprintRequest.Customizations.Users)[i].RedactSshKey()
+				}
+			}
```

Comment:
Okay, this is quite hard to read thanks to the wide ident. Let me first think about this part how I would write it; for each requestUser (`u`):

* Validate `u` first for invalid entries.
* Try to find existing user in the database using `slices.Contains` or `ContainsFunc`, store it as `ue`.
* If it does not exist, skip to the next user (no changes required).
* If it does exist, call the proposed `u.MergeUser(ue)` which will overwrite password/key if it is `nil`.
* Replace empty password/key (blank string) with `nil` (effectively deleting them).
* Validate `u` once again just in case the changes would create an invalid user.

Well, honestly it is tough to review your code it is quite complex. I suggest to clean it up, in general:

* Create a separate method which will accept the `User` type.
* Reuse the validation code.
* Try to explore the "merging" idea when there is no need to do prior checks, that should simplify it.
* Avoid creating hash to lookup existing records.
* And most importantly, write unit tests for the newly created function, it will be much easier to test a small method than testing the whole HTTP handler.